routing/bicycle_directions.cpp

```diff
@@ -0,0 +1,216 @@
+                                       my::Cancellable const & cancellable)
+{
+  size_t const pathSize = path.size();
+  CHECK_NOT_EQUAL(pathSize, 0, ());
+
+  times.clear();
+  turns.clear();
+  routeGeometry.clear();
+  m_adjacentEdges.clear();
+  m_pathSegments.clear();
+
+  auto emptyPathWorkaround = [&]()
+  {
+    turns.emplace_back(pathSize - 1, turns::TurnDirection::ReachedYourDestination);
+    this->m_adjacentEdges[0] = AdjacentEdges(1);  // There's one ingoing edge to the finish.
+  };
+
+  if (pathSize <= 1)
+  {
+    ASSERT(false, (pathSize));
+    emptyPathWorkaround();
+    return;
+  }
+
+  CalculateTimes(graph, path, times);
+
+  IRoadGraph::TEdgeVector routeEdges;
+  if (!ReconstructPath(graph, path, routeEdges, cancellable))
+  {
+    LOG(LDEBUG, ("Couldn't reconstruct path"));
+    emptyPathWorkaround();
+    return;
+  }
+  if (routeEdges.empty())
+  {
+    ASSERT(false, ());
+    emptyPathWorkaround();
+    return;
+  }
+
+  // Filling |m_adjacentEdges|.
+  m_adjacentEdges.insert(make_pair(0, AdjacentEdges(0)));
+  for (size_t i = 1; i < pathSize; ++i)
+  {
+    Junction const & prevJunction = path[i - 1];
+    Junction const & currJunction = path[i];
+    IRoadGraph::TEdgeVector outgoingEdges, ingoingEdges;
+    graph.GetOutgoingEdges(currJunction, outgoingEdges);
+    graph.GetIngoingEdges(currJunction, ingoingEdges);
+
+    AdjacentEdges adjacentEdges = AdjacentEdges(ingoingEdges.size());
+    // Outgoing edge angle is not used for bicyle routing.
+    adjacentEdges.m_outgoingTurns.isCandidatesAngleValid = false;
+    adjacentEdges.m_outgoingTurns.candidates.reserve(outgoingEdges.size());
+    ASSERT_EQUAL(routeEdges.size(), pathSize - 1, ());
+    FeatureID const inEdgeFeatureId = routeEdges[i - 1].GetFeatureId();
+
+    for (auto const & edge : outgoingEdges)
+    {
+      auto const & outFeatureId = edge.GetFeatureId();
+      // Checking for if |edge| is a fake edge.
+      if (!outFeatureId.IsValid())
+        continue;
+      adjacentEdges.m_outgoingTurns.candidates.emplace_back(0. /* angle */, outFeatureId.m_index,
+                                                            GetHighwayClass(outFeatureId));
+    }
+
+    LoadedPathSegment pathSegment;
+    if (inEdgeFeatureId.IsValid())
+      LoadPathGeometry(inEdgeFeatureId, {prevJunction.GetPoint(), currJunction.GetPoint()}, pathSegment);
+
+    m_adjacentEdges.insert(make_pair(inEdgeFeatureId.m_index, move(adjacentEdges)));
+    m_pathSegments.push_back(move(pathSegment));
+  }
+
+  RoutingResultGraph resultGraph(routeEdges, m_adjacentEdges, m_pathSegments);
+  RouterDelegate delegate;
+  Route::TTimes turnAnnotationTimes;
+  Route::TStreets streetNames;
+  MakeTurnAnnotation(resultGraph, delegate, routeGeometry, turns, turnAnnotationTimes, streetNames);
+}
+
+void BicycleDirectionsEngine::UpdateFeatureLoaderGuardIfNeeded(Index const & index, MwmSet::MwmId const & mwmId)
+{
+  if (!m_featuresLoaderGuard || mwmId != m_mwmIdFeaturesLoaderGuard)
```

Comment:
* FeaturesLoaderGuard has GetId() method, no need in m_mwmIdFeaturesLoaderGuard (in any case, I don't see a place where it's updated)
* as UpdateFeatureLoaderGuardIfNeeded() is always called before m_featuresLoaderGuard is used, why not to simplify things?
```c++
FeaturesLoaderGuard & GetLoader(MwmSet::MwmId const & id)
{
  if (!m_loader || id != m_loader->GetId())
    m_loader = make_unique<Index::FeaturesLoaderGuard>(m_index, id);
  return *m_loader;
}
```
Now, you don't need to pass Index parameter (as it's always m_index), and you can't forget to update FeaturesLoaderGuard if needed.